framework/resource_cache.h

```diff
@@ -80,82 +68,32 @@ class ResourceCache
 {
   public:
 	ResourceCache(Device &device);
-
-	ResourceCache(const ResourceCache &) = delete;
-
-	ResourceCache(ResourceCache &&) = delete;
-
+	ResourceCache(const ResourceCache &)            = delete;
+	ResourceCache(ResourceCache &&)                 = delete;
 	ResourceCache &operator=(const ResourceCache &) = delete;
+	ResourceCache &operator=(ResourceCache &&)      = delete;
 
-	ResourceCache &operator=(ResourceCache &&) = delete;
-
-	void warmup(const std::vector<uint8_t> &data);
-
-	std::vector<uint8_t> serialize();
-
-	void set_pipeline_cache(VkPipelineCache pipeline_cache);
-
-	ShaderModule &request_shader_module(VkShaderStageFlagBits stage, const ShaderSource &glsl_source, const ShaderVariant &shader_variant = {});
-
-	PipelineLayout &request_pipeline_layout(const std::vector<ShaderModule *> &shader_modules);
-
-	DescriptorSetLayout &request_descriptor_set_layout(const uint32_t                     set_index,
-	                                                   const std::vector<ShaderModule *> &shader_modules,
-	                                                   const std::vector<ShaderResource> &set_resources);
-
-	GraphicsPipeline &request_graphics_pipeline(PipelineState &pipeline_state);
-
-	ComputePipeline &request_compute_pipeline(PipelineState &pipeline_state);
-
-	DescriptorSet &request_descriptor_set(DescriptorSetLayout &                     descriptor_set_layout,
-	                                      const BindingMap<VkDescriptorBufferInfo> &buffer_infos,
-	                                      const BindingMap<VkDescriptorImageInfo> & image_infos);
-
-	RenderPass &request_render_pass(const std::vector<Attachment> &   attachments,
-	                                const std::vector<LoadStoreInfo> &load_store_infos,
-	                                const std::vector<SubpassInfo> &  subpasses);
-
-	Framebuffer &request_framebuffer(const RenderTarget &render_target,
-	                                 const RenderPass &  render_pass);
-
-	void clear_pipelines();
+	virtual ShaderModule        &request_shader_module(VkShaderStageFlagBits stage, const ShaderSource &glsl_source, const ShaderVariant &shader_variant = {});
```

Comment:
When you make all those functions virtual, you need to make the all the corresponding functions in HPPResourceCache virtual as well and have them in the same order to keep them congruent.
-> A loud call for some unifying approach between HPP and non-HPP again!

Reply:
Added virtual to missing functions and order is now the same as recommended. Would you be able to point me to some docs about this technique. I'd like to understand it more but couldn't find anything which explained it well

> A loud call for some unifying approach between HPP and non-HPP again!

We'll get there. I'm all for simplifying the framework and choosing a single version of the API to use for the majority. It very hard to make any useful change maintaining two versions of the same framework

Reply:
> Would you be able to point me to some docs about this technique.

I don't get what "technique" you mean here...
Currently, all the HPP- and corresponding non-HPP-classes are need to be binary identical, as they tend to be used in both HPP- and non-HPP calls and classes.
That is, all class/struct pairs need to have binary identical members in identical order, and apparently also need to have identical sets of virtual functions.

Reply:
This explains the issue i was having with the HPP sample. I should be able to test this at some point today